ENH: Improved `__str__` for polynomials

rossbar opened the pull request:
This PR follows up on some ideas originally introduced in #8893.

Since numpy v1.19 drops support for python2, one of the original ideas of #8893 can be revisited - using unicode sub/superscript values to prettify the printing of polynomials in `np.polynomial`.

This PR implements this feature by modifying the `__str__` method of the `ABCPolyBase` as well as introducing a few character mappings/helper methods to the `ABCPolyBase` class. 

Please let me know if this is a desired feature at this time.

Note that I modified the test suite to reflect the changes while trying to maintain some similarity to the original structure of the tests. I erred on the side of being explicit, though some reorganization of the tests could likely improve things. Also, I'm not sure if having unicode characters directly in the test suite will cause problems on some platforms vis-a-vis file encoding.

rossbar committed:
ENH: Improved __str__, __format__ for polynomials

Changes the printing style of instances of the convenience classes in
the polynomial package to a more "human-readable" format.
__str__ has been modified and __format__ added to ABCPolyBase, modifying
the string representation of polynomial instances, e.g. when printed.
__repr__ and the _repr_latex method (which is used in the Jupyter
environment are unchanged.

Two print formats have been added: 'unicode' and 'ascii'. 'unicode' is
the default mode on *nix systems, and uses unicode values for numeric
subscripts and superscripts in the polynomial expression. The 'ascii'
format is the default on Windows (due to font considerations) and uses
Python-style syntax to represent powers, e.g. x**2. The default
printing style can be controlled at the package-level with the
set_default_printstyle function.

The ABCPolyBase.__str__ has also been made to respect the linewidth
printoption. Other parameters from the printoptions dictionary are not
used.

Co-Authored-By: Warren Weckesser <warren.weckesser@gmail.com>
Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

charris commented:
Out of curiosity, how is line breaking handled?

rossbar commented:
As of right now, line breaking is not explicitly handled. That's certainly something that I would have to follow up on.

Also, after reading around some of the open issues/PRs surrounding `np.polynomial`, I would also propose adding info about the `domain` and `window` to  `__str__`. The default `__repr__` already includes this, but I think it should be included in `__str__` as well.

charris commented:
This will also need a release note.

eric-wieser commented on numpy/polynomial/_polybase.py:
```diff
@@ -67,6 +67,42 @@ class ABCPolyBase(abc.ABC):
+        6 : "\u2076",
+        7 : "\u2077",
+        8 : "\u2078",
+        9 : "\u2079"
```
Make the keys strings, and then you can use `str.translate` below

rossbar replied:
That's a great idea - better to use built-in functionality than reinvent the wheel. Addressed in 380c9f0d3

eric-wieser commented on numpy/polynomial/_polybase.py:
```diff
@@ -67,6 +67,42 @@ class ABCPolyBase(abc.ABC):
+        6 : "\u2086",
+        7 : "\u2087",
+        8 : "\u2088",
+        9 : "\u2089"
```
If you're not going to include the character directly, can you use the `\N` escapes which are more readable?

rossbar replied:
Good point - including the characters directly is probably the most readable, so I've made that change in 380c9f0d3

rossbar commented:
Thanks for the initial feedback @eric-wieser , those are definitely good ideas.

Re: linebreaks - it would be fairly straightforward to manually enforce 80-char limits and results in nice (e.g. no breaks in the middle of a single polynomial term), aligned output. However - introducing this manually here means that the printing wouldn't respect `np.printoptions` unless extra care was taken to do so. Do you know if there is any precedent for how other non-array classes in numpy behave wrt to `printoptions`?

eric-wieser commented on numpy/polynomial/_polybase.py:
```diff
@@ -284,9 +310,37 @@ def __repr__(self):
+                next_term += f"- {np.abs(coef)}"
+            # Handle different bases
+            if self.basis_name is None:
+                next_term += f"x{power.translate(self._superscript_mapping)} "
```
It's a bit weird to assume that a lack of basis_name implies a specific subclass.

It might be more consistent to define a `_str_term` to match the `_repr_latex_term`, and put the power series special case in the subclass override of that method

rossbar replied:
Good idea, I'll take a look.

rossbar replied:
I followed your suggestion in 795ceb6. I defined `_str_term` slightly differently than `_repr_latex_term`. since the `__str__` doesn't (yet) support the variable mapping between domain <-> window, I left off the `need_parens` argument.

rossbar commented:
To give a better sense of the effects of this PR, I'm including a few examples below to show the difference between the current string representation of polynomials, and how they look after this PR:

Currently:
```
>>> from numpy.polynomial import Polynomial, Chebyshev
>>> p, c = Polynomial([1, 2, 3]), Chebyshev([4, 5, 6])
>>> print(p)
poly([1. 2. 3.])
>>> print(c)
cheb([4. 5. 6.])
>>> print(p**2)
poly([ 1.  4. 10. 12.  9.])
>>> print(Polynomial([-1, 0, 0, 1]))
poly([-1.  0.  0.  1.])
>>> print(Chebyshev(np.arange(20)))
cheb([ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17.
 18. 19.])
```
With this PR:
```
>>> from numpy.polynomial import Polynomial, Chebyshev
>>> p, c = Polynomial([1, 2, 3]), Chebyshev([4, 5, 6])
>>> print(p)
1.0 + 2.0x¹ + 3.0x²
>>> print(c)
4.0 + 5.0T₁(x) + 6.0T₂(x)
>>> print(p**2)
1.0 + 4.0x¹ + 10.0x² + 12.0x³ + 9.0x⁴
>>> print(Polynomial([-1, 0, 0, 1]))
-1.0 + 0.0x¹ + 0.0x² + 1.0x³
>>> print(Chebyshev(np.arange(20)))
0.0 + 1.0T₁(x) + 2.0T₂(x) + 3.0T₃(x) + 4.0T₄(x) + 5.0T₅(x) + 6.0T₆(x) +
7.0T₇(x) + 8.0T₈(x) + 9.0T₉(x) + 10.0T₁₀(x) + 11.0T₁₁(x) + 12.0T₁₂(x) +
13.0T₁₃(x) + 14.0T₁₄(x) + 15.0T₁₅(x) + 16.0T₁₆(x) + 17.0T₁₇(x) + 18.0T₁₈(x) +
19.0T₁₉(x)
```

mattip commented:
This is an improvement, as the representation is more "mathematical" than "programatical". I wonder how this will affect useability for vision-impaired users.

Maybe it is my aging eyes, but the unicode sub/superscript values in some fonts seem to be very small. The [wikipedia page](https://en.wikipedia.org/wiki/Unicode_subscripts_and_superscripts) for these symbols explains that they were designed for fractions, and quotes the recommendation "When used in mathematical context (MathML) it is recommended to consistently use style markup for superscripts and subscripts...". When I copy-paste the rendered string into my terminal (which I uses Monospace Regular) the rendered string seems to be more readable than on this web page (using Deja Vu Sans). Scanning across the monospace fonts shows variation between them. It seems there is an opportunity for someone to tweak the existing fonts to render these symbols a little larger.

Note that `espeak "1.0 + 2.0x¹ + 3.0x²"` ignores the special formatting and says "one point zero plus two point zero x one plus three point zero x two".

I notice that IPython uses latex (I think, how do you choose between latex and html?) when outputting `display(p)`, due to [rich display] and the `_repr_latex_` method from gh-11528. 

The original issue gh-8893 mentions using `print("{:ascii}".format(p))`.  Maybe we could use that to render as `1.0 + 2.0x**1 + 3.0x**2`

rossbar commented:
> I wonder how this will affect useability for vision-impaired users.

This is a great point and something that I had not considered. I went the unicode route due to it's ubiquity in an attempt to come up with a solution that would be terminal-independent. It's doubly-unfortunate that `espeak` doesn't pick up the unicode super/subscripts. It would definitely be good to support the `:ascii` format option, which could hopefully mitigate some of these issues (though then there are other issues that would need to be addressed: discoverability of print options, how to set defaults, etc.). Another alternative is to use the `poly1d` approach, which uses full-sized ascii characters on a separate line:
```
   2
1 x + 2 x + 3
```
Although the font sizes are larger, in my opinion the readability suffers from the spacing. Any opinions/suggestions re: this issue would be very helpful.

> I notice that IPython uses latex (I think, how do you choose between latex and html?) when outputting display(p), due to rich display and the _repr_latex_ method from gh-11528.

Yes, LaTeX math is already used for the IPython front-ends that support it: namely Jupyter notebook and the QtConsole. As far as I know, most terminal emulators outside the frontends that are from the Jupyter project do not support these features, so the `_repr_latex_` (or any of the other `_repr_*_` methods) have no effect.

> The original issue gh-8893 mentions using print("{:ascii}".format(p)). Maybe we could use that to render as `1.0 + 2.0x**1 + 3.0x**2`

This is a really nice suggestion: I like the idea of defaulting to Python syntax for exponents when formatting with ascii. I think it would be also reasonable to use `_` for the other polynomials with different basis names, e.g. `1.0 + 2.0T_1(x) + 3.0T_2(x)` for Chebyshev series.

As mentioned above, maybe this mechanism can also help address the usability concerns for vision-impaired users.

eric-wieser commented:
I'm a little concerned that these characters are not going to work properly in a shell on windows.

I'm finding that I'm unable to even test if your code works, because copy-pasting your code into the windows terminal host strips all of the unicode characters. I suspect that it won't be able to display them either.

eric-wieser commented:
Digging further, the paste failing appears to be an ipython bug.

rossbar commented:
Thanks for the feedback. Platform-dependent encoding differences would definitely throw a wrench in the works. This is also a case where I'm not sure the tests will catch encoding differences.

rossbar commented:
Out of curiosity, do you see paste problems in IPython with the `%paste` command or just using the standard clipboard?

eric-wieser commented:
`%paste` appears to work correctly, thanks for reminding me why that existed.

Unfortunately depending on which of my open terminals I use and possibly the phase of the moon, I sometimes get

![image](https://user-images.githubusercontent.com/425260/77100473-9bd88980-6a0d-11ea-85ab-8f3c5ba22078.png)

That's probably a bug in either IPython, prompt-toolkit, or the terminal itself - but it's still possibly something we don't want to lead our users right into.

eric-wieser commented:
Ah, it's a font issue
- the default windows terminal raster font contains only the first 3 superscripts, reuses these for the subscripts,, and shows the remainder as regular numbers
- lucida console shows only the three superscripts, and boxes for everything else
- consolas is fine

charris commented:
Part of the reason I wished to leave such things to the display :) @eric-wieser Are there any standard display calls on any of the windows terminals that you know of? I know Jupyter has such things. I am not a fan of the ascii form.

IIRC, back when I was looked at this SymPy was recommended for having a good solution, it might be worth looking at how they do things and maybe stealing it. @asmeurer Thoughts?

asmeurer commented:
SymPy doesn't use unicode characters for superscript exponents. I think the font support for them isn't great, as you discovered. SymPy can also have arbitrary exponents, not just polynomials, which doesn't apply here. 

We do use `⋅` for multiplication

```
>>> pprint(2*x**2)
   2
2⋅x
```

As far as implementation, the SymPy printing system is far more generic than what would be needed here (it has to be able to print arbitrary symbolic expressions). Although you might steal the Unicode support detection code https://github.com/sympy/sympy/blob/c02515a8cfaa28c47300084b09b0d57798b867dd/sympy/printing/pretty/pretty_symbology.py#L50 (I don't know how it works, but I haven't heard any complaints about it).

rossbar commented:
Thanks for all the additional info! In terms of character support, I had been most worried about platform-dependent encoding issues, though that was addressed in [PEP 528](https://www.python.org/dev/peps/pep-0528/). 

I hadn't considered font issues however. As mentioned, it seems like that will be the main sticking point as it can't be counted on for the fonts used in various shells/terminals to support the range of unicode necessary for the full set of numerical super/subscripts.

> Although you might steal the Unicode support detection code https://github.com/sympy/sympy/blob/c02515a8cfaa28c47300084b09b0d57798b867dd/sympy/printing/pretty/pretty_symbology.py#L50 (I don't know how it works, but I haven't heard any complaints about it).

Thanks @asmeurer , this is a nice thing to be aware of. It looks like the checks in the linked function are limited to a small subset of specific unicode characters and doesn't probe the range for super/subscripts, but the concept could certainly be ported over.

The font consideration really seems to be the limiting factor for a generic approach that will be supported in all terminals. To me, the only viable options seem to be to stick to ascii either printing with Python syntax as mentioned above (i.e. `Polynomial([1, 2, 3])` looks like  `1.0 + 2.0x + 3.0x**2`) or to use the `np.poly1d`/`sympy` approach:
```
   2
3 x + 2 x + 1
```

It's a shame as, though it's not perfect, I found the unicode much more visually appealing.

rossbar commented:
**Summary**

Using unicode to represent super/subscripts for polynomial printing in the terminal is not a good approach due to font support issues and readability concerns. Two alternate approaches to Polynomial printing in the (non-Jupyter) terminal are presented below.

---

After the most recent round of revisions, I went back and reformulated things away from the unicode approach. I've implemented two options: "Python-style" formatting and "poly1d-style" printing (which, according to [the above discussion](https://github.com/numpy/numpy/pull/15666#issuecomment-601485085) appears to be consistent with `sympy`'s approach). The two styles are detailed below.

"Python-style" printing uses standard Python syntax for exponents and underscores to represent subscripts. Using similar examples [to those above](https://github.com/numpy/numpy/pull/15666#issuecomment-600971539), the printing would look something like:

```
>>> from numpy.polynomial import Polynomial, Chebyshev
>>> p, c = Polynomial([1, 2, 3]), Chebyshev([4, 5, 6])
>>> print(p)
1.0 + 2.0x**1 + 3.0x**2
>>> print(c)
4.0 + 5.0T_1(x) + 6.0T_2(x)
>>> print(p**2)
1.0 + 4.0x**1 + 10.0x**2 + 12.0x**3 + 9.0x**4
>>> print(Polynomial([-1, 0, 0, 1]))
-1.0 + 0.0x**1 + 0.0x**2 + 1.0x**3
>>> print(Polynomial(np.arange(10)))
0.0 + 1.0x**1 + 2.0x**2 + 3.0x**3 + 4.0x**4 + 5.0x**5 + 6.0x**6 + 7.0x**7 +
8.0x**8 + 9.0x**9
>>> print(Chebyshev(np.arange(10)))
0.0 + 1.0T_1(x) + 2.0T_2(x) + 3.0T_3(x) + 4.0T_4(x) + 5.0T_5(x) +
6.0T_6(x) + 7.0T_7(x) + 8.0T_8(x) + 9.0T_9(x)
```

"poly1d-style" printing uses only ascii characters, and denotes superscripts and subscripts on separate lines:

```
>>> from numpy.polynomial import Polynomial, Chebyshev
>>> p, c = Polynomial([1, 2, 3]), Chebyshev([4, 5, 6])
>>> print(p)
          1       2
1.0 + 2.0x  + 3.0x 

>>> print(c)
4.0 + 5.0T (x) + 6.0T (x)
          1          2   

>>> print(p**2)
          1        2        3       4
1.0 + 4.0x  + 10.0x  + 12.0x  + 9.0x 

>>> print(Polynomial([-1, 0, 0, 1]))
           1       2       3
-1.0 + 0.0x  + 0.0x  + 1.0x 

>>> print(Polynomial(np.arange(10)))
          1       2       3       4       5       6       7       8
0.0 + 1.0x  + 2.0x  + 3.0x  + 4.0x  + 5.0x  + 6.0x  + 7.0x  + 8.0x  + 

    9
9.0x 

>>> print(Chebyshev(np.arange(10)))
0.0 + 1.0T (x) + 2.0T (x) + 3.0T (x) + 4.0T (x) + 5.0T (x) + 6.0T (x) + 
          1          2          3          4          5          6   

7.0T (x) + 8.0T (x) + 9.0T (x)
    7          8          9   
```

To me, the "Python-style" printing has the advantage that it's more vertically compact, though it may be harder to interpret, especially for polynomials with many terms. The "poly1d-style" is nearer the traditional mathematical representation, but is not very compact vertically. One drawback of the "poly1d-style" is that it does not work well with non-monospace fonts (for example - take one of the printed outputs from the literal block above and paste it into an empty GitHub comment window - note that the alignment between the line containing the terms/coefficients and the line containing the powers will become misaligned).

Please let me know if you have thoughts/opinions about the relative merit of these two approaches. I like the "Python-style" best as it's the simplest and most portable, even though it's not very aesthetically pleasing. Another option is to leave the printing as-is and close the book on this particular attempt to update it. If you have thoughts, please chime in!

N.B. - Since there are two different options, I haven't modified the test suite, which is why it's currently failing. I will update it once a decision has been made.

stefanv commented:
My preference, for what it's worth, is the Python-like syntax, with spaces before the symbol:

```
0.0 + 1.0 x**1 + 2.0 x**2 + 3.0 x**3 + 4.0 x**4 + 5.0 x**5 + 6.0 x**6
+ 7.0 x**7 + 8.0 x**8 + 9.0 x**9
```

asmeurer commented:
The superscript numbers would probably be fine, so long as you detect if the terminal supports Unicode.

The detection itself is really about if the terminal supports Unicode encoding (i.e., you don't have `LANG=C` or something like that). It tells you nothing if the user will actually see the symbol, because that depends on if their computer has font support. I *think* the superscript numeric characters are pretty well supported, though it's something that you might want to verify. As I said, we primarily don't use them with SymPy because they don't extend well to general symbolic exponents. Actually, I think SymPy should consider using them in some cases https://github.com/sympy/sympy/issues/19049.

WarrenWeckesser commented:
@rossbar, here's an example where the linewidth logic doesn't quite work as expected:

```
In [26]: p = np.polynomial.Polynomial([np.pi, np.e, -9999, 1/3])

In [27]: np.set_printoptions(linewidth=41)

In [28]: print(p)
3.141592653589793 +
2.718281828459045x**1 - 9999.0x**2 + 0.3333333333333333x**3
```
The linewidth is set to 41, but that last line is 60 characters.

rossbar commented:
Thanks @WarrenWeckesser , I haven't updated the tests yet so there may well be other bugs/issues - that's definitely helpful!

rossbar commented:
@asmeurer I agree that font support is the main sticking point (along with readability concerns - the unicode super/subscripts are pretty small in most fonts).

> I think the superscript numeric characters are pretty well supported, though it's something that you might want to verify.

From [the wikipedia page](https://en.wikipedia.org/wiki/Unicode_subscripts_and_superscripts#Superscripts_and_subscripts_block) it looks like only 1, 2, and 3 were included in the Latin 1 character range, whereas the remaining super/subscripts are in their own dedicated block. From [the example above](https://github.com/numpy/numpy/pull/15666#issuecomment-601337138) it looks like the default font for some common shells include the Latin-1 range but not the extended unicode range required for general super/subscripts.

eric-wieser commented:
I suspect font support is a Windows-only issue. This post gives me some hope:

https://superuser.com/a/1398062/14403

Do OSX and Linux terminals use fallback fonts for missing characters? The Crux of the problem on Windows seems to be that it does not.

So maybe the behavior can be behind a platform detection switch.

anirudh2290 commented on numpy/polynomial/_polybase.py:
```diff
@@ -284,9 +284,41 @@ def __repr__(self):
+            next_term = ""
+            power = str(i + 1)
+            # Polynomial coefficient
+            if coef >= 0:
```
do we need to print 0 coef terms ? Is this to keep it consistent with latex representation ? my personal pref would be if its not blurred to omit it, since it may not be very readable in some cases for eg. sparse polynomials (as pointed out by @philer in https://github.com/numpy/numpy/issues/8893 )

rossbar replied:
No, the 0-coef terms don't need to be printed. I erred on the side of caution and decided to keep them in so as not to be "hiding" information from the user (after all, the 0's are part of the polynomial's `coef`).

Note that the `_repr_latex` keeps the zero terms, but "grays them out". This is a nice solution but I don't know if there is a comparable approach that could work for terminal output given that the support for rich output is so limited and platform/terminal dependent.

anirudh2290 commented on numpy/polynomial/_polybase.py:
```diff
@@ -284,9 +284,41 @@ def __repr__(self):
+            next_term += self._str_term(power, "x")
+            # Handle linebreaks
+            if (len(out) + len(next_term)) // linewidth != cur_line:
+                next_term = next_term.replace(" ", "\n", 1)
```
not sure if this will be interpreted as newline on windows or if it expects "\r\n"

rossbar replied:
I'm not sure - the test suite, which tests multiline printing, passes on the CI (I've reverted back to 7940d81f6 so that that the code is back in sync with the tests). Hopefully that is sufficient to indicate there aren't any problems expected w.r.t. newline characters.

eric-wieser replied:
This is handled by the stdout stream itself, no python code working with strings usually even needs to ever worry about this.

rossbar commented:
> Do OSX and Linux terminals use fallback fonts for missing characters? The Crux of the problem on Windows seems to be that it does not.

I'm not sure and it seems like behavior *could* vary quite a bit even on a single platform/distro depending on terminal emulators and their specific configuration.

> So maybe the behavior can be behind a platform detection switch.

Agreed - there could be multiple `_str_term` methods (one for ascii-only, and one using unicode sub/superscripts) with an internal toggle within `__str__` or the polybase class to determine which to use by default. This could also enable `:ascii` string formatting options.

Given the above discussion, I will do the following:
 1. Double-check the logic and add additional tests for handling linebreaks
 1. Include both the ascii and unicode approaches as separated `_str_x_term` methods.
 2. Include a toggle in the polybase class to set the default `_str_term` based on the platform
 3. map the ascii `_str_term` method to the `:ascii` formatting option.

Once all this is done, we can circle back and decide whether the unicode option is okay as a default (on non-Windows platforms). If the consensus at that point is "no", then the default can be adjusted accordingly.

Thank you everyone for the feedback!

eric-wieser commented on numpy/polynomial/_polybase.py:
```diff
@@ -283,10 +315,78 @@ def __repr__(self):
+            if coef >= 0:
+                next_term += f"+ {coef}·"
+            else:
+                next_term += f"- {np.abs(coef)}·"
```
Cdot is not ascii, this should be conditional on the formatting type too, using `*` for ascii.

rossbar replied:
Yes, this was intentional. The name `_use_unicode` that I chose for the switch is actually misleading, it should probably be `_use_unicode_superscripts`. The cdot is a unicode character, but it is in the Latin 1 range and thus it *should* be safe for all fonts.

I suppose though in the interest of a truly "failsafe" fallback it does make sense to remove all unicode characters. Thanks for the catch.

eric-wieser commented:
> My preference, for what it's worth, is the Python-like syntax, with spaces before the symbol:
> 
> ```
> 0.0 + 1.0 x**1 + 2.0 x**2 + 3.0 x**3 + 4.0 x**4 + 5.0 x**5 + 6.0 x**6 + 7.0 x**7 + 8.0 x**8 + 9.0 x**9
> ```

A possible alternative would be
```python
Polynomial.of(lambda x: 0.0 + 1.0*x**1 + 2.0*x**2 + 3.0*x**3 + 4.0*x**4 + 5.0*x**5 + 6.0*x**6 + 7.0*x**7 + 8.0*x**8 + 9.0*x**9)
```
and for the other types of polyomial:
```python
Chebyshev.of(lambda T, x: 0.0 + 1.0*T[0](x) + 2.0*T[2](x) + ...)
```
which has the nice property of actually being round-trippable back to a Polynomial instance.

rossbar commented:
I've incorporated the feedback since the [last large change](https://github.com/numpy/numpy/pull/15666#issuecomment-607000000) and implemented the plan roughly described [above](https://github.com/numpy/numpy/pull/15666#issuecomment-607609577). I'll try to give a quick summary here to prime the pump for the next round of reviewing.
 - I've implemented *both* printing styles, i.e. using unicode characters for superscripts and subscripts (and a cdot between the coefficients and terms), and an ascii-only format that using the Python-style printing suggested [in this comment](https://github.com/numpy/numpy/pull/15666#issuecomment-607385220). See the code box below the next bullet.
 - I've added a `__format__` method to `ABCPolyBase` that accepts the two formats, which I named 'unicode' and 'ascii':
    ```python
    >>> p = np.polynomial.Polynomial([1, 2, 3])
    >>> c = np.polynomial.Chebyshev([1, 2, 3])
    >>> print(f"{p:unicode}")
    1.0 + 2.0·x¹ + 3.0·x²
    >>> print(f"{c:unicode}")
    1.0 + 2.0·T₁(x) + 3.0·T₂(x)
    >>> print(f"{p:ascii}")
    1.0 + 2.0 x**1 + 3.0 x**2
    >>> print(f"{c:ascii}")
    1.0 + 2.0 T_1(x) + 3.0 T_2(x)
    ```
 - The default printing style is determined by the platform: \*nix platforms use 'unicode' by default while windows will use 'ascii'. This distinction reflects the discussion above about the default fonts that are common on different platforms and their support for the extended unicode character range necessary for all super/subscripts. Note that this simple approach is taken instead of a more in-depth approach based on testing the font support, and can easily be modified if the current platform-dependent default is undesirable.

In addition, in 56d5197df I added a "helper function" to the polynomial package (defined in `polynomial/__init__.py`) as an interface for changing the package-wide default for printing style. I'm not sure what the feeling is about adding functions like this to the package, though I think there should be some interface to allow users to easily change the printing style for the entire package (e.g. for users who find the unicode characters difficult to read).

WarrenWeckesser commented on numpy/polynomial/tests/test_printing.py:
```diff
@@ -1,38 +1,217 @@
-    def test_legendre_str(self):
-        res = str(poly.Legendre([0, 1]))
-        tgt = 'leg([0. 1.])'
+class TestStrAscii:
```
PEP 8: Two blank lines before `class`.

WarrenWeckesser commented on numpy/polynomial/tests/test_printing.py:
```diff
@@ -1,38 +1,217 @@
 import numpy.polynomial as poly
 from numpy.testing import assert_equal
 
+class TestStrUnicodeSuperSubscripts:
```
PEP 8: Two blank lines before `class`.

WarrenWeckesser reviewed (commented):
I made a bunch of nit-picky comments in-line--the only substantial one is the question about the argument to `__format__` being required.

Another issue I noticed is that there is a small off-by-something error in the line-breaking code.  For example, the ascii string for `p = np.polynomial.Polynomial([123456789, 123456789, 123456789, 1234, 1])` should fit in 74 characters (i.e. `'123456789.0 + 123456789.0 x**1 + 123456789.0 x**2 + 1234.0 x**3 + 1.0 x**4'`).  I have the linewidth set to the default 75, but the current code puts in a line break.

Finally, experience with style settings like this has shown that eventually someone will want to use a context manager for it.  We don't need it now, and maybe we never will, so we probably shouldn't worry about it, as long as the implementation details (such as the `_use_unicode` attribute attached to each class) are private, so if there is ever demand for a context manager, it can be added without requiring a backwards incompatible change. It looks like that is the case here.

WarrenWeckesser commented on numpy/polynomial/_polybase.py:
```diff
@@ -283,10 +315,78 @@ def __repr__(self):
         name = self.__class__.__name__
         return f"{name}({coef}, domain={domain}, window={window})"
 
+    def __format__(self, fmt_str):
```
Should the argument really be *required*?  Why not default to the `__str__` behavior?  As it is, for a polynomial `p` such as `p = np.polynomial.Polynomial([1, 2, 3])`, both `format(p)` and `f'{p}'` raise an exception. E.g.

```
In [138]: p = np.polynomial.Polynomial([1, 2, 3])

In [139]: f'{p}'
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-139-8e0bd8161cf8> in <module>
----> 1 f'{p}'

~/mc37np/lib/python3.7/site-packages/numpy-1.19.0.dev0+56d5197-py3.7-macosx-10.9-x86_64.egg/numpy/polynomial/_polybase.py in __format__(self, fmt_str)
    319         if fmt_str not in ('ascii', 'unicode'):
    320             raise TypeError(
--> 321                 f"Unsupported format string '{fmt_str}' passed to "
    322                 f"{self.__class__}.__format__. Valid options are "
    323                 f"'ascii' and 'unicode'"

TypeError: Unsupported format string '' passed to <class 'numpy.polynomial.polynomial.Polynomial'>.__format__. Valid options are 'ascii' and 'unicode'
```

WarrenWeckesser replied:
Or rather, when the argument is `''`, use the same logic as `__str__`.

rossbar commented:
Thanks for the review @WarrenWeckesser, your comments are extremely helpful. 

> Should the argument really be required? Why not default to the __str__ behavior?

Great catch - this was an oversight on my part. I agree with your comment as to what the behavior should be and will fix accordingly.

> Another issue I noticed is that there is a small off-by-something error in the line-breaking code. 

I thought I had chased these issues down, but as you mention it looks like I'm off a bit on the character count. I will add a +/- 1 whitespace test and fix whatever issue remains here.

The in-line comments look good to - thanks for the review!

rossbar commented on doc/release/upcoming_changes/15666.improvement.rst:
```diff
@@ -0,0 +1,7 @@
+--------------------------------------------------------
+
+The string representation (``__str__``)of all six polynomial types in
+`numpy.polynomial` now use unicode characters to represent powers
```
Good catch, I completely missed it. I'll probably have to revisit the release notes again to add something about the two different str representations (ascii and Unicode).

rossbar commented on numpy/polynomial/_polybase.py:
```diff
@@ -283,10 +315,78 @@ def __repr__(self):
+        if linewidth < 1:
+            linewidth = 1
+        out = ""
+        if self.coef[0] < 0:
```
Nice catch, thanks!

rossbar commented on numpy/polynomial/_polybase.py:
```diff
@@ -283,10 +315,78 @@ def __repr__(self):
+        out = f"{self.coef[0]} "
+        cur_line = len(out) // linewidth    # For handling linebreaks
+        for i, coef in enumerate(self.coef[1:]):
+            next_term = ""
```
Good point, the line initializing `next_term` is unnecessary.

rossbar commented:
The latest set of commits (ending in dd9048e) incorporate the feedback from [the most recent review](https://github.com/numpy/numpy/pull/15666#pullrequestreview-396686533) (thanks again @WarrenWeckesser!). Changes include:

 - Several style changes in _polybase and the test suite to better abide by PEP 8
 - The default case for an empty formatting string (as you get in `f"{p}"` or `format(p)`) now defaults to `__str__`, as suggested.
 - The line breaking issue in `__str__` has been fixed. The tests for testing linebreaks have been replaced with more precise tests to evaluate string expressions with `len` +/- from the desired linewidth.
 - Various other small changes to remove unnecessary variable initializations and conditionals
 - Release note has been updated to reflect some of the new work since it was originally added.

I had marked the conversations above as resolved to keep track of the changes as I made them - if something is not sufficiently addressed please feel free to re-open them.

WarrenWeckesser commented on numpy/polynomial/tests/test_printing.py:
```diff
@@ -1,39 +1,262 @@
+            '123456789.0 + 123456789.0 x**1 + 123456789.0 x**2 + '
+            '1234.0 x**3 + 1.0 x**4'
+        ))
+    
```
Remove trailing whitespace.
```suggestion

```

WarrenWeckesser commented on numpy/polynomial/__init__.py:
```diff
@@ -20,6 +20,54 @@
 from .hermite_e import HermiteE
 from .laguerre import Laguerre
 
+def set_default_printstyle(style):
```
PEP 8: Two blank lines before `def`.
```suggestion

def set_default_printstyle(style):
```

WarrenWeckesser commented on numpy/polynomial/__init__.py:
```diff
@@ -20,6 +20,54 @@
+
+    Notes
+    -----
+    The default format depends on the platform: 'unicode' is used on 
```
Remove trailing whitespace.
```suggestion
    The default format depends on the platform: 'unicode' is used on
```

WarrenWeckesser commented on numpy/polynomial/__init__.py:
```diff
@@ -20,6 +20,54 @@
+        _use_unicode = False
+    from ._polybase import ABCPolyBase
+    ABCPolyBase._use_unicode = _use_unicode
+
```
`flake8` wants two blank lines here, too.
```suggestion


```

WarrenWeckesser commented on numpy/polynomial/_polybase.py:
```diff
@@ -67,6 +68,37 @@ class ABCPolyBase(abc.ABC):
 
+    # Unicode character mappings for improved __str__
+    _superscript_mapping = str.maketrans({
+        "0" : "⁰",
```
You have a space before the `:` in these dictionary items.  I always thought that was against the PEP 8 recommendations, but after taking another look at the document, I see that it never explicitly says that. However, the examples in the PEP 8 document don't use a space before `:` when creating a dictionary item, and `flake8` will complain about it, so I recommend not using a space.   But I won't insist, and I don't know if anyone else will mind, so your call...

rossbar replied:
I'm unaware of what PEP8 dictates here either, but if it's getting caught by linters then I agree it should probably be changed - thanks!

WarrenWeckesser commented on numpy/polynomial/_polybase.py:
```diff
@@ -67,6 +68,37 @@ class ABCPolyBase(abc.ABC):
+    })
+    # Some fonts don't support full unicode character ranges necessary for
+    # the full set of superscripts and subscripts, including common/default
+    # fonts in Windows shells/terminals. Therefore, default to ascii-only 
```
Trailing whitespace
```suggestion
    # fonts in Windows shells/terminals. Therefore, default to ascii-only
```

WarrenWeckesser commented on numpy/polynomial/_polybase.py:
```diff
@@ -283,10 +315,82 @@ def __repr__(self):
+        if fmt_str == '':
+            return self.__str__()
+        if fmt_str not in ('ascii', 'unicode'):
+            raise TypeError(
```
Also should be `ValueError`.  Python uses `ValueError`, e.g.
```
In [14]: f"{100:q}"
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-14-f0e76aaf133d> in <module>
----> 1 f"{100:q}"

ValueError: Unknown format code 'q' for object of type 'int'

In [15]: f"{100:ABC}"
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-15-188fbe0b055e> in <module>
----> 1 f"{100:ABC}"

ValueError: Invalid format specifier
```

rossbar replied:
Nice catch, thanks for pointing this out!

WarrenWeckesser commented on numpy/polynomial/_polybase.py:
```diff
@@ -283,10 +315,82 @@ def __repr__(self):
+            next_term += term_method(power, "x")
+            # Length of the current line with next term added
+            line_len = len(out.split('\n')[-1]) + len(next_term)
+            # If not the last term in the polynomial, it will be two 
```
Trailing whitespace
```suggestion
            # If not the last term in the polynomial, it will be two
```

WarrenWeckesser commented on numpy/polynomial/_polybase.py:
```diff
@@ -283,10 +315,82 @@ def __repr__(self):
+    @classmethod
+    def _str_term_unicode(cls, i, arg_str):
+        """
+        String representation of single polynomial term using unicode 
```
Trailing whitespace
```suggestion
        String representation of single polynomial term using unicode
```

WarrenWeckesser reviewed (commented):
Looks good.  I made a bunch more nit-picky whitespace suggestions, plus one API suggestion to replace a `TypeError` with `ValueError` when the format string is not `'unicode'` or `'ascii'`.

rossbar commented on numpy/polynomial/__init__.py:
```diff
@@ -20,6 +20,54 @@
+    1.0 + 2.0·x¹ + 3.0·x²
+    """
+    if style not in ('unicode', 'ascii'):
+        raise TypeError(
```
:+1:

rossbar commented:
Thanks so much @WarrenWeckesser for another close look - your suggestions should be reflected in f8688cb:
 * Removed the whitespace preceding the `:` in the dictionary definitions
 * `TypeError` -> `ValueError` in the two string formatting functions

Also - sorry for not linting before pushing, it will go on the list of pre-PR tasks for the future!

WarrenWeckesser reviewed (approved):
Looks good.  I think the history can be squashed to just one or two commits.  @rossbar, can you rebase with a squash?  That way you can edit the commit message(s) to include just the essential information, without the churn.  (No need to preserve my commit authorship for the "suggested edits".)

rossbar commented:
Will do @WarrenWeckesser , thank you (and others) for the careful review!

rossbar commented:
I've squashed everything down into acdaff7 and written a new commit message to give an overview of the changes (and preserving co-authorship, assuming I've done that correctly).

eric-wieser commented on numpy/polynomial/_polybase.py:
```diff
@@ -283,10 +315,82 @@ def __repr__(self):
+            if coef >= 0:
+                next_term = f"+ {coef}"
+            else:
+                next_term = f"- {np.abs(coef)}"
```
```suggestion
                next_term = f"- {-coef)}"
```
would be faster

rossbar replied:
Good point - @seberg also pointed out that `abs` (and potentially the `>=0` comparison) would cause problems with complex coefficients. I will add a test for complex coefficients as well.

eric-wieser commented on numpy/polynomial/_polybase.py:
```diff
@@ -67,6 +68,37 @@ class ABCPolyBase(abc.ABC):
+    # the full set of superscripts and subscripts, including common/default
+    # fonts in Windows shells/terminals. Therefore, default to ascii-only
+    # printing on windows.
+    _use_unicode = not os.name == 'nt'
```
Should this be thread-local?

seberg replied:
As long as not even print options seem to be, not sure its worth worrying about?

eric-wieser replied:
That's fair.

rossbar commented:
Okay, 9c83b13 incorporates @eric-wieser 's suggestion and adds a test to verify that complex coefficients are printed correctly. I've re-squashed everything into 9c83b13 and (hopefully) maintained attribution correctly.

I admit I don't entirely understand the implications in the thread-local discussion, so I'll leave the conversation unresolved in case someone wants to expand on it.

seberg commented:
Do we have to do more about `x >= 0` possibly failing? I realize its fine for numpy complex numbers...

As to the threads, it is about things such changing the setting in a subthread modifying the global state. If this was a context manager (or used like one), then surprising things can happen unless you make it thread-local.  I think we should not worry about it. Although, it might mean that we need to turn this into a property or a global getter function at some point.
I suppose subclasses (made by users) in theory could use this underscored item. Although, though should be fine if we ever turn it into a property...

rossbar commented:
> Do we have to do more about x >= 0 possibly failing? I realize its fine for numpy complex numbers...

This is a good point. In principle, the coefficient arrays can be `object` arrays, which means that it is possible to construct polynomials with non-numeric coefficients (even if this doesn't make sense):

```python
>>> p = np.polynomial.Polynomial(np.array([1, 'f', 2.0], dtype=object))
>>> p
Polynomial([1, 'f', 2], dtype=object, domain=[-1,  1], window=[-1,  1])
>>> print(p)
TypeError: '>=' not supported between instances of 'str' and 'int'
```

Since this currently allowed, then `__str__` should work in this case as well. I'd be inclined to wrap `_generate_string()` in a `try/except` that falls back to the `__repr__`.

rossbar committed:
Handle TypeError in _generate_str for coefs.

Add a fallback for TypeErrors that are raised when attempting
to compare arbitrary elements (e.g. strings or Python complex)
to 0 in _generate_str.

rossbar commented:
f923959 wraps `_generate_string()` in a try/except that catches `TypeError`s from failed `>=` comparisons between incompatible types. In that case, `__str__` falls back to `__repr__`, which simply gives the array of coefficients rather than trying to construct a nice polynomial expression from them.

I've also updated the tests to verify that this fallback works, as well as test `object` arrays that contain "numerical" objects like `Decimal`s or `Fraction`s.

rossbar commented:
I've reworked how `_generate_str` handles coefficients for which `>= 0` raises an exception (e.g. strings or Python complex). A try/except catches TypeError exceptions from this comparison and falls back to simply representing the coefficient in its current form.

I've also added tests for the Python complex case.

rossbar committed:
STY: Move comment outside of try/except.

seberg commented:
@WarrenWeckesser I am happy with the local try/except solution and IIRC the rest was pretty far along. But I never reviewed this in detail. Do you want to do the honors?

rossbar committed:
DOC: Update poly class refguide printing.

Update routines.polynomials.classes doc in the refguide to reflect
changes to polynomial printing.

Add additional information to the document about the various ways that
the string representation of polynomial expressions can be controlled
via formatting.

WarrenWeckesser commented:
@rossbar, the change in doc/source/reference/routines.polynomials.classes.rst looks good.

@charris or anyone else: do you have any objections to merging this so close to the branching of 1.19?

mattip commented:
Thanks @rossbar
